text/0046-metadata-for-offline-signers.md

```diff
@@ -0,0 +1,274 @@
+
+#### Reduce metadata size
+
+Metadata should be stripped from parts that are not necessary to parse a signable extrinsic, then it should be separated into a finite set of self-descriptive chunks. Thus, a subset of chunks necessary for signable extrinsic decoding and rendering could be sent, possibly in small portions (ultimately - one at a time), to cold device together with proof.
+
+ - Single chunk with proof payload size SHOULD fit within few kB;
+ - Chunks handling mechanism SHOULD support chunks being sent in any order without memory utilization overhead;
+ - Unused enum variants MUST be stripped (this has great impact on transmitted metadata size; examples: era enum, enum with all calls for call batching).
+
+## Stakeholders
+
+All chain teams are stakeholders, as implementing this feature would require timely effort on their side and would impact compatibility with older tools.
+
+This feature is essential for **all** offline signer tools; many regular signing tools might make use of it. In general, this RFC greatly improves security of any network implementing it, as many governing keys are used with offline signers.
+
+Implementing this RFC would remove requirement to maintain metadata portals manually, as task of metadata verification would be effectively moved to consensus mechanism of the chain.
+
+## Explanation
+
+Detailed description of metadata shortening and digest process is provided in [metadata-shortener](https://github.com/Alzymologist/metadata-shortener) crate (see `cargo doc --open` and examples). Below are presented algorithms of the process.
+
+### Definitions
+
+#### Metadata structure
+
+Metadata in general consists of four sections:
+
+1. Types registry
+2. Pallets
+3. Extrinsic metadata
+4. Runtime type
+
+Of these, only sections 1-3 contain information required for extrinsic decoding. The most important section is (1) Types registry, that is mostly used in extrinsic decoding. It is also the largest part, thus it is modularized for fractional transmission. Part (2) contains runtime version and is otherwise useless for transaction decoding; thus its contents are reduced to this parameter and included into Metadata Descriptor. Part (3) is included into Metadata Descriptor verbatim.
+
+#### Metadata descriptor
+
+Values for:
+
+1. `u8` metadata shortening protocol version, 
+2. SCALE-encoded `ExtrinsicMetadata`,
+3. SCALE-encoded `spec_version` `String`,
+4. SCALE-encoded `spec_name` `String`,
+5. `u16` base58 prefix,
+6. `u8` decimals value or `0u8` if no units are defined,
+7. SCALE-encoded `tokenSymbol` `String` defined on chain to identify the name of currency (available for example through `system.properties()` RPC call) or empty string if no base units are defined,
+
+```
+struct MetadataDescriptor { // really a scale-encoded enum, thus first field is enum value - only 0x01 currently supported.
+  protocol_version: u8,
```

Comment:
AFAIK the byte that we have in the digest is the same as this one. We're then 100% sure that both in runtime build and in metadata shortening process (that happen in different envs and different times) the same process was used.

It seems to be the `index` of the enum actually. Changing the declaration to an enum would simplify the understanding of it.
I'd as well mention that `V0` is already burned.

Reply:
> AFAIK the byte that we have in the digest is the same as this one. We're then 100% sure that both in runtime build and in metadata shortening process (that happen in different envs and different times) the same process was used.

But we already put this byte into the tx and it will be signed. So, we already ensure this?

Reply:
Yeh, true... I still see value for the devices parsing the metadata (since it's scale, they need to know the structure). With that being an enum and not a simple struct, we're sure that we are parsing the info correctly (in terms of metadata, not blob), and it also gives us the flexibility to change it in any chain and keep backwards-compatibility as well with the previous versions.
Otherwise, we need the blob to know how to parse the metadata, and we need the metadata to know how to parse the blob. It's kind of a vicious circle.

Reply:
It's just one byte, but a byte that makes implementations a bit more ideomatic and helps developers explicitly state which version of shortener protocol they are using at the moment in their code. I think it's worth it. It's not even mandatory to send it every time - signer device could add it itself - if sending one extra byte matters.